File: src/lib/util/string_checker.ts
import {LinkedList} from "./linked_list";

const BRANCH_NUM_MAX = 256;

class Node {
    depth: number = 0;
    next: { [char: number]: Node } = {};
    fail: Node = null;
}

export class StringChecker {
    root: Node = new Node();

    addPattern(pattern: string) {
        let buffer = new Buffer(pattern);
        let depth = 0;
        let node = this.root;
        for (let i = 0; i < buffer.length; i++) {
            ++depth;

            let k = buffer[i];
            if (k >= 'A'.charCodeAt(0) && k <= 'Z'.charCodeAt(0)) {
                k = k - 'A'.charCodeAt(0) + 'a'.charCodeAt(0);
            }

            if (!node.next[k]) {
                node.next[k] = new Node();
            }
            node = node.next[k];
        }
        node.depth = depth;
    }

    build() {
        let q: LinkedList<Node> = new LinkedList<Node>();
        this.root.fail = null;
        q.append(this.root);

        while (q.length > 0) {
            let node = q.head;
            q.deleteNode(node);
            for (let i = 0; i < BRANCH_NUM_MAX; i++) {
                if (!node.element.next[i]) {
                    continue;
                }
                let p = node.element.fail;
                while (p) {
                    if (p.next[i]) {
                        node.element.next[i].fail = p.next[i];
                        break;
                    }
                    p = p.fail;
                }
                if (p === null) {
                    node.element.next[i].fail = this.root;
                }
                q.append(node.element.next[i]);
            }
        }
    }

    replace(src: string): string {
        let buffer = new Buffer(src);
        let tmp = '';
        let node = this.root;
        let st = 0;
        let ed = 0;

        while (ed !== buffer.length) {
            let k = buffer[ed];
            if (k >= 'A'.charCodeAt(0) && k <= 'Z'.charCodeAt(0)) {
                k = k - 'A'.charCodeAt(0) + 'a'.charCodeAt(0);
            }

            while (!node.next[k] && node !== this.root) {
                node = node.fail;
            }

            if (node.next[k]) {
                node = node.next[k];
            }

            if (node !== this.root && node.depth) {
                let data = new Buffer(ed - node.depth + 1 - st);
                buffer.copy(data, 0, st, ed - node.depth + 1);
                tmp += data.toString();
                tmp += '***';
                node = this.root;
                st = ed + 1;
            }
            ++ed;
        }

        if (st !== src.length) {
            let data = new Buffer(ed - st);
            buffer.copy(data, 0, st, ed);
            tmp += data.toString();
        }

        return tmp;
    }

    reset() {
        this.root = new Node();
    }
}

File: src/lib/addons/stringchecker/StringChecker.cpp
#include "StringChecker.h"

StringChecker::StringChecker()
{
	root = new Node();
}

StringChecker::~StringChecker()
{
	DeleteTree( root );
}

void StringChecker::AddPattern( const char* pat )
{
	uint16 	depth 	= 0;
	Node*	node	= root;
	while ( *pat )
	{
		++depth;

		uint8 k = *pat;

		// all lower case
		if ( k >= 'A' && k <= 'Z' ) k = k - 'A' + 'a';

		if ( node->next[k] == NULL )
		{
			node->next[k] = new Node();
		}
		node = node->next[k];
		++pat;
	}
	node->depth = depth;
}

void StringChecker::Build()
{
	std::queue< Node* > q;

	root->fail = NULL;
	q.push( root );

	while ( !q.empty() )
	{
		Node* node = q.front(); q.pop();
		for ( int i = 0; i < BRANCH_NUM_MAX; i++ )
		{
//			if ( node->next[i] == NULL ) continue;
			if ( node->next.find(i) == node->next.end() ) continue;
			Node* p = node->fail;
			while ( p )
			{
//				if ( p->next[i] )
				if(p->next.find(i) != p->next.end())
				{
					node->next[i]->fail = p->next[i];
					break;
				}
				p = p->fail;
			}
			if ( p == NULL )
			{
				node->next[i]->fail = root;
			}
			q.push( node->next[i] );
		}
	}
}

bool StringChecker::Replace( std::string& src )
{
	bool ret = false;
	std::string tmp;
	std::string::iterator st = src.begin();
	std::string::iterator ed = src.begin();

	Node* node = root;

	while ( ed != src.end() )
	{
		uint8 k =  *ed; // it must be cast to Unsigned !!

		// all lower case
		if ( k >= 'A' && k <= 'Z' ) k = k - 'A' + 'a';

//		while ( node->next[k] == NULL && node != root )
		while( node->next.find(k) == node->next.end() && node != root)
		{
			node = node->fail;
		}

		if (  node->next.find(k) != node->next.end() ) node = node->next[k];
//		if ( node->next[k] ) node = node->next[k];

		if ( node != root && node->depth )
		{
			tmp.append( st, ed - node->depth + 1 );
			tmp.append( 3, '*' ); // replace

			node = root;// restart

			st = ed + 1;
			ret = true;
		}
		++ed;
	}

	if ( st != src.end() )
	{
		tmp.append( st, ed );
	}

	src = tmp;

	return ret;
}

void StringChecker::Reset()
{
	DeleteTree( root );
	root = new Node();
}

void StringChecker::DeleteTree( Node* node )
{
	if(node == NULL) return ;
	for ( std::map<uint8,Node*>::iterator iter = node->next.begin(); iter != node->next.end(); iter++)
	{
//		if ( node->next[i] ) DeleteTree( node->next[i] );
		DeleteTree( iter->second );
	}
	delete node;
}
